sError/sUndefined doesn't eta expand, leading to "Defined but not used" warnings with -Wall

This program emits a warning:

```haskell
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
{-# OPTIONS_GHC -Wall -ddump-splices #-}
module Bug where

import Data.Singletons.Prelude
import Data.Singletons.TH

$(singletons
    [d| f :: Bool -> Bool
        f = error "f"
      |])
```

```
GHCi, version 8.2.1: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/rgscott/.ghci
[1 of 1] Compiling Bug              ( Bug.hs, interpreted )
Bug.hs:(16,3)-(19,8): Splicing declarations
    singletons
      [d| f_a3Mj :: Bool -> Bool
          f_a3Mj = error "f" |]
  ======>
    f_a6rK :: Bool -> Bool
    f_a6rK = error "f"
    type FSym1 (t_a6rN :: Bool) = F t_a6rN
    instance SuppressUnusedWarnings FSym0 where
      suppressUnusedWarnings _
        = snd ((GHC.Tuple.(,) FSym0KindInference) GHC.Tuple.())
    data FSym0 (l_a6rO :: TyFun Bool Bool)
      = forall arg_a6rP. SameKind (Apply FSym0 arg_a6rP) (FSym1 arg_a6rP) =>
        FSym0KindInference
    type instance Apply FSym0 l_a6rO = F l_a6rO
    type family F (a_a6rQ :: Bool) :: Bool where
      F a_6989586621679034577_a6rR = Apply (Apply ErrorSym0 "f") a_6989586621679034577_a6rR
    sF ::
      forall (t_a6rS :: Bool).
      Sing t_a6rS -> Sing (Apply FSym0 t_a6rS :: Bool)
    sF (sA_6989586621679034577 :: Sing a_6989586621679034577_a6rR)
      = sError (sing :: Sing "f")

Bug.hs:16:3: warning: [-Wunused-matches]
    Defined but not used: ‘sA_6989586621679034577’
   |
16 | $(singletons
   |   ^^^^^^^^^^...
```

The `-ddump-splices` output should make it apparent as to why: `sF` is binding a variable `sA_6989586621679034577` but never using it. Curiously, this only happens for the singled version of `f`, as the promoted `F` _does_ do eta expansion correctly!

--- a/src/Data/Singletons/Single.hs
+++ b/src/Data/Singletons/Single.hs
@@ -551,9 +551,9 @@ singExp (ADAppE e1 e2) _res_ki = do
   e1' <- singExp e1 Nothing
   e2' <- singExp e2 Nothing
   -- `applySing undefined x` kills type inference, because GHC can't figure
-  -- out the type of `undefined`. So we don't emit that code.
+  -- out the type of `undefined`. So we don't emit `applySing` there.
   if isException e1'
-  then return e1'
+  then return $ e1' `DAppE` e2'
   else return $ (DVarE applySingName) `DAppE` e1' `DAppE` e2'
 singExp (ADLamE ty_names prom_lam names exp) _res_ki = do
   let sNames = map singValName names

--- a/tests/compile-and-dump/Singletons/T167.ghc82.template
+++ b/tests/compile-and-dump/Singletons/T167.ghc82.template
@@ -131,7 +131,7 @@ Singletons/T167.hs:(0,0)-(0,0): Splicing declarations
       sFooList
         (sA_0123456789876543210 :: Sing a_0123456789876543210)
         (sA_0123456789876543210 :: Sing a_0123456789876543210)
-        = sUndefined
+        = (sUndefined sA_0123456789876543210) sA_0123456789876543210
     instance SFoo a => SFoo [a] where
       sFoosPrec ::
         forall (t :: Nat) (t :: [a]) (t :: [Bool]).

--- a/tests/compile-and-dump/Singletons/Undef.ghc82.template
+++ b/tests/compile-and-dump/Singletons/Undef.ghc82.template
@@ -34,6 +34,6 @@ Singletons/Undef.hs:(0,0)-(0,0): Splicing declarations
     sFoo ::
       forall (t :: Bool). Sing t -> Sing (Apply FooSym0 t :: Bool)
     sBar (sA_0123456789876543210 :: Sing a_0123456789876543210)
-      = sError (sing :: Sing "urk")
+      = (sError (sing :: Sing "urk")) sA_0123456789876543210
     sFoo (sA_0123456789876543210 :: Sing a_0123456789876543210)
-      = sUndefined
+      = sUndefined sA_0123456789876543210

--- a/tests/compile-and-dump/Singletons/Undef.hs
+++ b/tests/compile-and-dump/Singletons/Undef.hs
@@ -1,7 +1,7 @@
+{-# OPTIONS_GHC -Wall #-}
 module Singletons.Undef where
 
 import Data.Singletons.TH
-import Data.Singletons.Prelude
 
 $(singletons [d|
   foo :: Bool -> Bool

